design-documents/graph-ql/mutation-error-design.md

```diff
@@ -0,0 +1,153 @@
+# Mutation Error Design
+
+Mutations are _the_ single way to trigger side-effects in a GraphQL API. Side-effects are where the large majority of our errors are going to happen, as a result of changing some underlying data.
```

Comment:
Please see https://graphql.org/learn/queries/#mutations . I believe "side effect" is usually referring to unexpected effect like data modification during query.

```suggestion
Mutations should be _the_ single way used to modify server-side data in a GraphQL API. The large majority of our errors are going to happen during mutations, as a result of changing some underlying data.
```

Reply:
The language used is from the GraphQL specification's description of mutations:

> Mutation
If the operation is a mutation, the result of the operation is the result of executing the mutation’s top level selection set on the mutation root object type. This selection set should be executed serially.

> It is expected that the top level fields in a mutation operation perform side‐effects on the underlying data system. Serial execution of the provided mutations ensures against race conditions during these side‐effects.

http://spec.graphql.org/June2018/

Reply:
Still not convinced. The intention, the expected effect of the mutation is state modification. Why do we call it side-effect in this case? 

Which of these (from the wikipedia) cases do we have?
```
Example side effects include modifying a non-local variable, modifying a static local variable, modifying a mutable argument passed by reference, performing I/O or calling other side-effect functions.
```

Reply:
> an operation, function or expression is said to have a side effect if it modifies some state variable value(s) outside its local environment, that is to say has an observable effect besides returning a value (the main effect) to the invoker of the operation

Note that the terminology of side-effects is also used in the HTTP specification when describing "safe methods." We're just sharing the common terminology both of these specifications use.

Reply:
Agree, that side-effect is usually referred to any state modifications. It's different from the same non-science term (which has a negative connotation).

Reply:
I replaced usage of the term "side-effect" with something that should be clearer 😃